[Multi-Screen Window Placement API ] Update article for second OT

tomayac opened the pull request:
Updated the article for the second OT.

Do not merge before the OT URL is up.

quisquous commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -63,8 +61,12 @@ enable the `#enable-experimental-web-platform-features` flag in `about://flags`.
+some [changes](https://github.com/webscreens/window-placement/blob/main/CHANGES.md) to the API. The
+article has been updated accordingly.
+
+Starting in Chromium&nbsp;92, the Multi-Screen Window Placement API will again be available as an
+origin trial in Chromium. This second origin trial is expected to end in Chromium&nbsp;95.
```
```suggestion
Starting in Chromium&nbsp;93, the Multi-Screen Window Placement API will again be available as an
origin trial in Chromium. This second origin trial is expected to end in Chromium&nbsp;96.
```

We had a discussion today and decided to push this back to M93.

quisquous commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
-that resolves with an array of `Screen` objects. On my MacBook Air 13 with a connected iPad, this
-returns an array of two `Screen` objects:
+Now that I know that the current setup is multi-screen, I can obtain more information about the
+second screen using `Window.getScreens()`. It returns a promise that resolves with an array of
```
Is it worth mentioning that this prompts the user for permission here?

michaelwasserman replied:
That is covered under "The window-placement permission", but this might be an okay place to remind readers of that.

tomayac replied:
I'll add it.

quisquous commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
````diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
+});
+```
+
+### The `currentscreenchange` event
````
Thanks for the text on all the new events.  These look good to me!

tomayac replied:
:-) Thanks for making the events more explicit, this will help developers subscribe to just the things they need info on.

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
````diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
 ```
 
 The browser
 [can](https://webscreens.github.io/window-placement/#issue-c601b517:~:text=Permission%20check%20for%20above%3F,-partial)
````
Please update this link, maybe https://webscreens.github.io/window-placement/#usage-overview-screen-information:~:text=This%20method%20may%20prompt%20the%20user%20for%20permission

tomayac replied:
Done.

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
-new or an existing screen is (physically or virtually in the case of Sidecar) plugged in or unplugged.
+`screenschange`, does exactly that: it fires whenever the screen constellation is modified. (Notice
+that "screens" is plural in the event name.) It also fires when the resolution of one of the
+connected screens changes or when a new or an existing screen is (physically or virtually in the
```
`screenschange` does *not* fire when attributes of a particular screen change, only when some screens are added or removed from the `screens` array. For that, users should use the new `change` event on `Screen` (and `ScreenAdvanced`). Let us know if we can clarify https://webscreens.github.io/window-placement/#api-screens-events or the explainer, etc.

tomayac replied:
Thanks, clarified.

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
-future](https://github.com/webscreens/window-placement/issues/28). For now you
-can look up the screen details by calling `window.getScreens()` as shown below.
+itself does not provide this data. This may
+[change in the future](https://github.com/webscreens/window-placement/issues/28). For now you can
```
I closed that issue; we won't add data to the events themselves. Please let us know if you have remaining ergonomics issues/suggestions regarding events, thanks!

tomayac replied:
With the live objects and the caching of the initial call result of `getScreens()` it still feels like the events passed to the event listeners are a bit superfluous, but it's definitely workable.

michaelwasserman replied:
Is it possible and acceptable API design to trigger event listeners without passing an event object? If so, maybe that would be desirable here?

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
-can look up the screen details by calling `window.getScreens()` as shown below.
+itself does not provide this data. This may
+[change in the future](https://github.com/webscreens/window-placement/issues/28). For now you can
+look up the screen details by calling `window.getScreens()` as shown below.
```
They can, but really shouldn't need to call getScreens() again. Instead, they can/should use the *live* multi-screen data of a cached `Screens` interface object.

tomayac replied:
Understood.

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
-  console.log('I am there, but mostly useless', event);
+const screensInterface = await window.getScreens();
+screensInterface.addEventListener('screenschange', async (event) => {
   const details = await window.getScreens();
```
Consider avoiding the repeated call to getScreens. One practical example here might be something like:

const screensInterface = await window.getScreens();
let cachedScreensLength = screensInterface.screens.length;
screensInterface.addEventListener('screenschange', event => {
  if (screensInterface.screens.length != cachedScreensLength) {
    console.log('The screen count changed from' + cachedScreensLength + ' to' screensInterface.screens.length);
    cachedScreensLength = screensInterface.screens.length;
  }
});

Generally, `screenschange` should correspond to a change in the length of `screens`, but technically it's possible that a single event might fire if N screens were added and N screens were removed simultaneously...

tomayac replied:
Noted and changed.

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
````diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
+```js
+const screensInterface = await window.getScreens();
+screensInterface.addEventListener('currentscreenchange', async (event) => {
+  const details = (await window.getScreens()).currentScreen;
````
Similarly, reuse screensInterface instead of calling getScreens() again

tomayac replied:
Done.

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
````diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
+```js
+const firstScreen = (await window.getScreens())[0];
+firstScreen.addEventListener('change', async (event) => {
+  const details = (await window.getScreens())[0];
````
ditto

tomayac replied:
Done.

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -292,29 +345,29 @@ try {
 
-The other aspects of the API—the `onscreenschange` event and the `screen` property of the
-`FullscreenOptions`—would simply never fire or silently be ignored respectively by non-supporting
+The other aspects of the API, that is, the `screenschange` event and the `screen` property of the
```
nit: "that is, the screen change events and" (there are multiple events that wouldn't fire)

tomayac replied:
Gotcha.

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -425,10 +478,8 @@ The Multi-Screen Window Placement API spec was edited by
 [Victor Costan](https://www.linkedin.com/in/pwnall) and
 [Joshua Bell](https://www.linkedin.com/in/joshuaseanbell). The API was implemented by
 [Mike Wasserman](https://www.linkedin.com/in/mike-wasserman-9900a079/). This article was reviewed by
```
Let's credit enne with API development too!

tomayac replied:
Done.

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
````diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
 ```
 
 The browser
 [can](https://webscreens.github.io/window-placement/#issue-c601b517:~:text=Permission%20check%20for%20above%3F,-partial)
````
```suggestion
[can](https://webscreens.github.io/window-placement/#usage-overview-screen-information:~:text=This%20method%20may%20prompt%20the%20user%20for%20permission)
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
-that resolves with an array of `Screen` objects. On my MacBook Air 13 with a connected iPad, this
-returns an array of two `Screen` objects:
+Now that I know that the current setup is multi-screen, I can obtain more information about the
+second screen using `Window.getScreens()`. It returns a promise that resolves with an array of
```
```suggestion
second screen using `Window.getScreens()`. Calling this function will show a permission prompt that asks me whether the site may open and place windows on my screen. The function returns a promise that resolves with an array of
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
-It also fires when the resolution of one of the connected screens changes or when a
-new or an existing screen is (physically or virtually in the case of Sidecar) plugged in or unplugged.
+`screenschange`, does exactly that: it fires whenever the screen constellation is modified. (Notice
+that "screens" is plural in the event name.) It also fires when the resolution of one of the
+connected screens changes or when a new or an existing screen is (physically or virtually in the
+case of Sidecar) plugged in or unplugged.
```
```suggestion
that "screens" is plural in the event name.) This means the event fires whenever a new or an existing screen is (physically or virtually in the
case of Sidecar) plugged in or unplugged.
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
-itself does not provide this data. This may [change in the
-future](https://github.com/webscreens/window-placement/issues/28). For now you
-can look up the screen details by calling `window.getScreens()` as shown below.
+itself does not provide this data. This may
+[change in the future](https://github.com/webscreens/window-placement/issues/28). For now you can
+look up the screen details by calling `window.getScreens()` as shown below.
```
```suggestion
itself does not provide this data. To look up the screen details, use the live object from a cached `Screens` interface.
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
````diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
 ```js
-window.addEventListener('screenschange', async (event) => {
-  console.log('I am there, but mostly useless', event);
+const screensInterface = await window.getScreens();
+screensInterface.addEventListener('screenschange', async (event) => {
   const details = await window.getScreens();
+  console.log('The screen constellation has changed.', event), details;
+});
````
```suggestion
const screensInterface = await window.getScreens();
let cachedScreensLength = screensInterface.screens.length;
screensInterface.addEventListener('screenschange', (event) => {
  if (screensInterface.screens.length !== cachedScreensLength) {
    console.log(`The screen count changed from ${cachedScreensLength} to ${screensInterface.screens.length}`);
    cachedScreensLength = screensInterface.screens.length;
  }
});
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
````diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
+`currentScreen`), I can listen for the `currentscreenchange` event.
+
+```js
+const screensInterface = await window.getScreens();
+screensInterface.addEventListener('currentscreenchange', async (event) => {
+  const details = (await window.getScreens()).currentScreen;
+  console.log('The current screen has changed.', event, details);
+});
````
```suggestion
const screensInterface = await window.getScreens();
screensInterface.addEventListener('currentscreenchange', async (event) => {
  const details = screensInterface.currentScreen;
  console.log('The current screen has changed.', event, details);
});
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
````diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
+```js
+const firstScreen = (await window.getScreens())[0];
+firstScreen.addEventListener('change', async (event) => {
+  const details = (await window.getScreens())[0];
````
```suggestion
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,154 @@ The new `window-placement` permission can be queried with the
+const firstScreen = (await window.getScreens())[0];
+firstScreen.addEventListener('change', async (event) => {
+  const details = (await window.getScreens())[0];
+  console.log('The first screen has changed.', event, details);
```
```suggestion
  console.log('The first screen has changed.', event, firstScreen);
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -292,29 +345,29 @@ try {
 
-The other aspects of the API—the `onscreenschange` event and the `screen` property of the
-`FullscreenOptions`—would simply never fire or silently be ignored respectively by non-supporting
+The other aspects of the API, that is, the `screenschange` event and the `screen` property of the
```
```suggestion
The other aspects of the API, that is, the various screen change events and the `screen` property of the
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -425,10 +478,8 @@ The Multi-Screen Window Placement API spec was edited by
 [Victor Costan](https://www.linkedin.com/in/pwnall) and
 [Joshua Bell](https://www.linkedin.com/in/joshuaseanbell). The API was implemented by
 [Mike Wasserman](https://www.linkedin.com/in/mike-wasserman-9900a079/). This article was reviewed by
```
```suggestion
[Mike Wasserman](https://www.linkedin.com/in/mike-wasserman-9900a079/) and [Adrienne Walker](https://github.com/quisquous). This article was reviewed by
```

tomayac commented:
Thanks, @quisquous and @michaelwasserman. I have addressed your remarks and comments :-) PTAL!

Not sure who, but either of you needs to register their GitHub account with Google (https://opensource.corp.google.com/github/) so the Google CLA bot is happy (it's [grumpy](https://github.com/GoogleChrome/web.dev/pull/5591#issuecomment-859447895) right now).  After that, please comment `@googlebot I fixed it.`.

michaelwasserman commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -338,9 +394,9 @@ You can play with the [demo][demo] embedded below, or see its [source code][demo
 <!-- Copy and Paste Me -->
 <div class="glitch-embed-wrap" style="height: 800px; width: 100%;">
-  <iframe src="https://glitch.com/embed/#!/embed/window-placement?path=iframe.html&previewSize=100"
+  <iframe src="https://window-placement.glitch.me/"
```
The feature now has permission policy support, so if an iframe will actually attempt to use the API, it will the iframe will need an "allow='window-placement'" attribute.

michaelwasserman replied:
Hmm, clicking elmer just opens the demo in a separate tab, so perhaps we don't need to support the iframe itself using getScreens(), etc.

tomayac replied:
I have (had) logic in the demo that would recognize when it was iframed and open in a new tab. With the new permission policy integration we can get rid of this.

michaelwasserman commented:
Thanks Tom! The updated article lgtm, modulo the potential iframe permission policy issue.

Thanks for updating the demo too, but I think it might need a bit of additional polish: Clicking a popup on the primary/internal screen on Mac enters and immediately exits fullscreen, then closes and reopens all popups. I'm guessing screenschange / currentscreenchange / screen.change is fired when entering/exiting fullscreen. This could be because my mac's gpu switches (http://crbug.com/1218952) or just because the current screen metrics (e.g. the avail* dimensions) may rightly change on fullscreen toggle. Hopefully we can refine the demo a bit to avoid super glitchy cases; lmk if you'd like some help with that.

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -338,9 +394,9 @@ You can play with the [demo][demo] embedded below, or see its [source code][demo
+  <iframe src="https://window-placement.glitch.me/"
     title="window-placement on Glitch"
-    allow="fullscreen; geolocation; microphone; camera; midi; vr; encrypted-media"
+    sandbox="allow-popups allow-popups-to-escape-sandbox allow-scripts"
```
```suggestion
    allow="window-placement" sandbox="allow-popups allow-popups-to-escape-sandbox allow-scripts"
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -7,26 +7,24 @@ description:
   machine and to place windows on specific screens.
 date: 2020-09-14
-updated: 2021-02-23
+updated: 2021-06-11
```
```suggestion
updated: 2021-06-14
```

tomayac replied:
```suggestion
updated: 2021-06-14
```

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -338,9 +394,9 @@ You can play with the [demo][demo] embedded below, or see its [source code][demo
+  <iframe src="https://window-placement.glitch.me/"
     title="window-placement on Glitch"
-    allow="fullscreen; geolocation; microphone; camera; midi; vr; encrypted-media"
+    allow="window-placement" sandbox="allow-popups allow-popups-to-escape-sandbox allow-scripts"
```
```suggestion
    allow="fullscreen; window-placement" sandbox="allow-modals allow-popups allow-popups-to-escape-sandbox allow-scripts"
```

tomayac commented:
Thanks, @michaelwasserman! I have now modified the listener for `currentscreenchange` to be a no-op (apart from logging), since it always fires whenever fullscreen is entered. I tried detecting this with `document.fullscreen`, but it wouldn't work. We can revisit this when http://crbug.com/1218952 gets fixed.

jpmedley commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -7,26 +7,24 @@ description:
   machine and to place windows on specific screens.
 date: 2020-09-14
-updated: 2021-02-23
+updated: 2021-06-14
```
Do you know when the OT will be live? We could use `scheduled: true` and merge this as soon as the edit is done. (I prefer that because nobody has to remember to loop back and merge this.)

tomayac replied:
@michaelwasserman will know.

michaelwasserman replied:
We are pretty confident we'll start OT2 in M93. Maybe set the date around the first M93 Beta (Thu July 29)?

tomayac replied:
@michaelwasserman, if you ping the URL once you know it, I can then put it into the article and take care of merging if @jpmedley approves the article in the current state, with the only pending addition being the URL (and updating the publishing date to whatever date it may be).

tomayac replied:
I have added the `updated` date that is in the far future :-) of July and the `scheduled: true` flag. Not sure if this works for already published articles, but with the `Do not merge` label we're also on the safe side.

jpmedley commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,157 @@ The new `window-placement` permission can be queried with the
-of the methods of the new API.
-Read on to learn more.
+[can](https://webscreens.github.io/window-placement/#usage-overview-screen-information:~:text=This%20method%20may%20prompt%20the%20user%20for%20permission)
+choose to show the permission prompt dynamically upon the first attempt to use any of the methods of
```
```suggestion
choose to show the permission prompt dynamically on the first attempt to use any of the methods of
```

jpmedley commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,157 @@ The new `window-placement` permission can be queried with the
+allows me to determine exactly how the screens are arranged logically (next to each other, on top of
+each other, etc.). There is also data now for each screen to show whether it is an `isInternal` one
+and whether it is an `isPrimary` one. Note that the built-in screen
 [is not necessarily the primary screen](https://osxdaily.com/2010/04/27/set-the-primary-display-mac/#:~:text=Click%20on%20the%20Display%20icon,primary%20display%20for%20your%20Mac).
```
Isn' necessarily or doesn't have to be? The former implies an element of chance, while the later implies my complete control.

tomayac replied:
"Isn't necessarily". Operating systems may have different heuristics as to which plugged in additional screen they make the primary screen.

jpmedley commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -168,106 +171,157 @@ The new `window-placement` permission can be queried with the
+arrangements to be restored. Information about the possible input methods are available in the
+`pointerTypes` array.
+
+The field `currentScreen` is a live object corresponding to the current `window.screen`. The object
```
```suggestion
The `currentScreen` field is a live object corresponding to the current `window.screen`. The object
```

quisquous commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -7,26 +7,25 @@ description:
 alt: Simulated trading desk showing multiple fake cryptocurrencies and their price charts.
 origin_trial:
-  url: https://developers.chrome.com/origintrials/#/view_trial/1411878483180650497
+  url: https://developers.chrome.com/origintrials
```
```suggestion
  url: https://developer.chrome.com/origintrials/#/view_trial/-8087339030850568191
```

Here's the new origin trial link.

tomayac commented on src/site/content/en/blog/multi-screen-window-placement/index.md:
```diff
@@ -49,7 +47,7 @@ Examples of sites that may use this API include:
 | 2. Create initial draft of specification | [Complete][spec]         |
 | 3. Gather feedback & iterate on design   | [In progress](#feedback) |
-| **4. Origin trial**                      | **[In progress][ot]**    |
+| 4. Origin trial                          | Upcoming                 |
```
```suggestion
| 4. **Origin trial**                          | **[In progress](https://developer.chrome.com/origintrials/#/view_trial/-8087339030850568191)**                 |
```
